### DISCUSSION TITLE
Select different PHP Versions in different Projects

### DISCUSSION THREAD

**Discussion by kazanove** [leokhoa/laragon#416] [Q&A] [answered]
how to install multiple versions  PHP?

**Comment by Pen-y-Fan** [comment]
See my blog post: [How to update the PHP version in Laragon](https://pen-y-fan.github.io/2023/01/15/how-to-update-the-php-version-in-laragon/)

**Comment by KenyettaYuilli5m** [comment]
To install multiple versions of PHP on your system, you can follow these general steps:

Add the repositories for the PHP versions you want to install to your package manager. For example, on Ubuntu, you can add the ondrej/php repository to install different versions of PHP.

Install the desired PHP versions using your package manager. For example, to install PHP 7.4 and PHP 8.0 on Ubuntu, you can run the following commands:
![image](https://user-images.githubusercontent.com/127737344/224670056-4bc8a73a-7a30-415c-aff0-f277475dde40.png)
After installing multiple PHP versions, you can switch between them using a tool like update-alternatives. For example, to switch to PHP 8.0, you can run:
![image](https://user-images.githubusercontent.com/127737344/224670133-b238f9b7-c783-4ca4-b331-c7cc954c6636.png)
And to switch back to PHP 7.4, you can run:
![image](https://user-images.githubusercontent.com/127737344/224670210-6ff66965-20d4-4f1b-b3c9-be9a618cbcea.png)
You can also configure different PHP versions for different web servers or virtual hosts by modifying their configuration files to use the appropriate PHP executable.

Note that the specific steps for installing multiple versions of PHP may vary depending on your operating system and package manager.

**Comment by TemekaRomanellichm** [comment]
To install multiple versions of PHP, you can use a tool like phpbrew or multiple PHP-FPM instances.

**Comment by HasanAlyazidi** [comment]
Is there any way to use a different PHP version for each project using `.htaccess` or something?

**Reply by RobiNN1** [reply]
It's not a great solution, but it works. I used it for a while

Example with php 7.0 and 7.3.

Add this to the end of your apache config (E.g. `C:/laragon/bin/apache/httpd-2.4.41-win64-VC14/conf/httpd.conf`) before `IncludeOptional ...`.

```
ScriptAlias /php70 "C:/laragon/bin/php/php-7.0.33-Win32-VC14-x64"
Action application/x-httpd-php70-cgi /php70/php-cgi.exe
<Directory "C:/laragon/bin/php/php-7.0.33-Win32-VC14-x64">
    AllowOverride None
    Options None
    Require all denied
    <Files "php-cgi.exe">
        Require all granted
    </Files>
</Directory>

ScriptAlias /php73 "C:/laragon/bin/php/php-7.3.33-Win32-VC14-x64"
Action application/x-httpd-php73-cgi /php73/php-cgi.exe
<Directory "C:/laragon/bin/php/php-7.3.33-Win32-VC14-x64">
    AllowOverride None
    Options None
    Require all denied
    <Files "php-cgi.exe">
        Require all granted
    </Files>
</Directory>
```

And then in project's .htaccess set this `SetHandler application/x-httpd-php70-cgi` or `SetHandler application/x-httpd-php73-cgi`

**Reply by tomerof** [reply]
how do i do it in nginx ?
i have two projects that i need to run simultaneously, one is php 7.3.3 and the other 8.3.7
is there a way to tell nginx to use version 8.3.7 while the main laragon version is 7.3.3 for apache.

**Reply by asif-ca** [reply, marked as answer]
Amazingly, this worked for me! However, images were not loading correctly, especially .svg files, which failed to appear on the page.

The issue occurred because, after setting `SetHandler application/x-httpd-php70-cgi`, every file was being served by CGI as PHP.

I resolved this by configuring the CGI handler to apply only to PHP files in .htaccess file:

# Handle PHP files using this specific CGI handler
```
<FilesMatch "\.php$">
    SetHandler application/x-httpd-php70-cgi
</FilesMatch>
```

Now even laragon has php 8 by defualt in my case but my projects can use php7 (or any specified version by sethandler)  without switching php version in laragon.

**Comment by Pen-y-Fan** [comment]
How to manually **pin** Apache to a set version of PHP.

Thanks to [RobiNN1](https://github.com/RobiNN1) for his reply above and [SO: Setting PHP Include Path on a per site basis?](https://stackoverflow.com/questions/24622/setting-php-include-path-on-a-per-site-basis) I have found a way to manually pin one site to a fixed PHP version, while Laragon can still switch PHP to another version and Apache will automatically use Laragon's PHP.

There are some caveats:

1. The manual site must use an NTS version of PHP so Apache will use **mod_fcgid**
2. The manual site needs to have a different TLD
3. The Windows host file needs to be manually updated
4. The PHP version set in Laragon must be a TS version of PHP so Apache will use **mod_php**.
5. Only one PHP version can be pinned at a time.

Download and install the required NTS version of PHP see: [How to update the PHP version in Laragon](https://pen-y-fan.github.io/2023/01/15/how-to-update-the-php-version-in-laragon/), if the PHP version is 'legacy' see the [archive](https://windows.php.net/downloads/releases/archives/)

Update the Windows host file with a different TLD than Laragon. e.g. **app.stranding.local**

```text
127.0.0.1	app.stranding.local # manually set
127.0.0.1      stranding.test       #laragon magic! 
````

Create an Apache conf file for the site: e.g. manual.app.stranding.local.conf in **sites_enabled** (Laragon menu > Apache > dir: conf) double click sites-enabled

```text
define ROOT "C:/laragon/www/stranding/public"
define SITE "app.stranding.local"
define PHPPATH "C:/laragon/bin/php/php-8.2.3-nts-Win32-vs16-x64"
LoadModule fcgid_module "C:/laragon/etc/apache2/modules/mod_fcgid-2.3.10-win64-VS16.so"

<IfModule fcgid_module>
FcgidInitialEnv PATH "${PHPPATH};C:/Windows/system32;C:/Windows;C:/Windows/System32/Wbem;"
FcgidInitialEnv SystemRoot "C:/Windows"
FcgidInitialEnv SystemDrive "C:"
FcgidInitialEnv TEMP "C:/Windows/Temp"
FcgidInitialEnv TMP "C:/Windows/Temp"
FcgidInitialEnv windir "C:/Windows"

# 10 hrs: in case you have long running scripts, increase FcgidIOTimeout 
FcgidIOTimeout 36000
FcgidConnectTimeout 16
FcgidMaxRequestsPerProcess 0
FcgidMaxProcesses 50
FcgidMaxRequestLen 81310720
# Location php.ini:
FcgidInitialEnv PHPRC "${PHPPATH}"
FcgidInitialEnv PHP_FCGI_MAX_REQUESTS 0


</IfModule>

<VirtualHost *:80> 

    DocumentRoot "${ROOT}"
    ServerName ${SITE}
    ServerAlias *.${SITE}
    <Directory "${ROOT}">
        AllowOverride All
        Require all granted
    </Directory>
	<Files ~ "\.php$>"
		SetHandler fcgid-script
		Options +ExecCGI
		FcgidWrapper "${PHPPATH}/php-cgi.exe" .php
	</Files>

</VirtualHost>

<VirtualHost *:443>

    DocumentRoot "${ROOT}"
    ServerName ${SITE}
    ServerAlias *.${SITE}
    <Directory "${ROOT}">
        AllowOverride All
        Require all granted
    </Directory>
	<Files ~ "\.php$>"
		SetHandler fcgid-script
		Options +ExecCGI
		FcgidWrapper "${PHPPATH}/php-cgi.exe" .php
	</Files>
	
	SSLEngine on
    SSLCertificateFile      C:/laragon/etc/ssl/laragon.crt
    SSLCertificateKeyFile   C:/laragon/etc/ssl/laragon.key

</VirtualHost>

```

Edit the file paths and names in the first four lines:

1. _define ROOT "C:/laragon/www/stranding/public"_ is the entry point into your app
2. _define SITE "app.stranding.local"_ is the manual name set in the Windows host file
3. _define PHPPATH "C:/laragon/bin/php/php-8.2.3-nts-Win32-vs16-x64"_ is the PHP version you want to fix
4. _LoadModule fcgid_module "C:/laragon/etc/apache2/modules/mod_fcgid-2.3.10-win64-VS16.so"_  is the Laragon Apache version for mod_fcgid, it may not need updating if you are using the Laragon 6 default Apache.

⚠️ All the paths should be forward slash / not the windows backslash \

Laragon can be restarted, the site will be pinned. You can now switch to a different PHP version in Laragon, for other sites.

![image](https://github.com/leokhoa/laragon/assets/40126936/08c693d5-93b3-494b-9fa0-f258e95a5ef9)


![image](https://github.com/leokhoa/laragon/assets/40126936/5ac9690b-688d-49e1-ad9b-b4c8c2b4d5c0)


When you have finished pinning the PHP version either rename the file so it doesn't end .conf or delete it.

I hope this helps.

**Reply by Ringoinmel** [reply]
Hi Pen-y-fan
Any idea?  I wonder this line FcgidInitialEnv PHPRC "${PHPPATH}" doesn't work.  Thanks for your help

**Reply by Ringoinmel** [reply]
**Let me explain it into more details.**
Default php = php-8.1.10-Win32-vs16-x64 

**code in the file**
define ROOT "C:/laragon/www"
define SITE "http://lmsmaster.freeddns.org"
define PHPPATH "C:/laragon/bin/php/php-7.4.33-nts-Win32-vc15-x64"

**result**
phpinfo shows PHP Version 7.4.33 on the top left corner
Loaded Configuration File = C:\laragon\bin\php\php-8.1.10-Win32-vs16-x64\php.ini
extension_dir = C:/laragon/bin/php/php-8.1.10-Win32-vs16-x64/ext

**Reply by Pen-y-Fan** [reply]
The first two lines don't look right to me.

```ini
define ROOT "C:/laragon/www/stranding/public" is the entry point into your app
define SITE "app.stranding.local" is the manual name set in the Windows host file
```

Your local project maybe in **C:/laragon/www/lmsmaster**, the entry point (ROOT) might be **C:/laragon/www/lmsmaster/public**

define SITE would then be **lmsmaster.local** and it should have the same entry in **hosts**

**Comment by Ringoinmel** [comment]


**Reply by Pen-y-Fan** [reply]
As long as the pinned version uses NTS, in this example PHP 7.4 is the NTS version, and Laragon is using PHP 8.1 TS version, it works fine for me. If I set PHP in Laragon to use a NTS version, the configuration file will not work properly. Remember to Reload Apache after making any changes to .conf files too.

![image](https://github.com/user-attachments/assets/7b0378c5-0b09-4ef1-9f0e-8c6599e6ee3c)

**Reply by Pen-y-Fan** [reply]
```ini
define ROOT "F:/laragon/www/stranding/public"
define SITE "app.stranding.local"
define PHPPATH "F:/laragon/bin/php/php-7.4.33-nts-Win32-vc15-x64"
```

**Reply by Pen-y-Fan** [reply]
![image](https://github.com/user-attachments/assets/488b4c62-2636-434f-805f-ca61d41b11c1)

**Comment by Ringoinmel** [comment]
It's working now.  You are champion.  Thanks a lot.

I think the problem I have is I should not have 2 copy of php in TS.  Once I delete the other one, it works.  Both websites in 2 different php version can run at the same time.

Thanks Bro

**Comment by izmeez** [comment]
@RobiNN1 Your suggestion in https://github.com/leokhoa/laragon/discussions/416#discussioncomment-6999723 works with Apache 2.4.62.x but not with 2.4.65.x

How does it need to be changed to work with Apache 2.4.65.x ?
